lib/adagrams.rb

```diff
@@ -0,0 +1,161 @@
+# adds csv gem
+require "csv"
+
+# initialize global variable
+$words_hash = {}
+
+# wave 1
+def draw_letters
+  # draw letters
+
+  # letter pool
+  pool = ('A'*9 + 'B'*2 + 'C'*2 + 'D'*4 + 'E'*12 + 'F'*2 + 'G'*3 + 'H'*2 + 'I'*9 + 'J' + 'K' + 'L'*4 + 'M'*2 + 'N'*6 + 'O'*8 + 'P'*2 + 'Q' + 'R'*6 + 'S'*4 + 'T'*6 + 'U'*4 + 'V'*2 + 'W'*2 + 'X' + 'Y'*2 + 'Z').chars
+
+  # randomize letters
+  # return the first 10 letters from randomized pool
+  return pool.shuffle.pop(10)
+end
+
+# wave 2
+def uses_available_letters?(input, letters)
+  # input --> user's word (string)
+  # letters --> available letters in pool
+
+  # made a clone of letters; previously, we ran into memory location/duplication issues
+  # we don't want to alter letters since it needs to reset with all 10 pulled letters available upon each call
+  letters_copy = letters.clone
```

Comment:
We'll talk about reference vs. value in class formally soon. But also, I'm pretty curious, what side effects did you see if you didn't make a copy of `letters`?

Reply:
If you input in an incorrect word, and then on the second try input in a valid word using any of the same letters, the program will not recognize it as valid because it modifies the hand of letters after the first attempt.

Reply:
If I'm remembering correctly, we had the following issue if we didn't make a copy: if someone drew the letters "a", "b", "c", "d", "e", "f" for example and made the nonsense word "abc" then they would be asked to try again. But "a", "b", and "c" would have been mistakenly deleted from the list, leaving only "d", "e", and "f". So if they next tried the word "cab" they would get an error message. We also tried letters_copy = letters, but I think we figured out that they're pointing to the same place in memory because the "copy" would have the same issue.

Reply:
Right! Got it, makes sense. Thanks!